nwb_conversion_tools/datainterfaces/icephys/baseicephysinterface.py

```diff
@@ -0,0 +1,145 @@
+        metadata: dict = None,
+        stub_test: bool = False,
+        use_times: bool = False,
+        save_path: OptionalFilePathType = None,
+        overwrite: bool = False,
+        write_as: str = "raw",
+        es_key: str = None,
+        icephys_experiment_type: Optional[str] = None,
+    ):
+        """
+        Primary function for converting raw (unprocessed) RecordingExtractor data to the NWB standard.
+
+        Parameters
+        ----------
+        nwbfile: NWBFile
+            nwb file to which the recording information is to be added
+        metadata: dict
+            metadata info for constructing the nwb file (optional).
+            Should be of the format
+                metadata['Ecephys']['ElectricalSeries'] = dict(name=my_name, description=my_description)
+        use_times: bool
+            If True, the times are saved to the nwb file using recording.frame_to_time(). If False (default),
+            the sampling rate is used.
+        save_path: PathType
+            Required if an nwbfile is not passed. Must be the path to the nwbfile
+            being appended, otherwise one is created and written.
+        overwrite: bool
+            If using save_path, whether or not to overwrite the NWBFile if it already exists.
+        stub_test: bool, optional (default False)
+            If True, will truncate the data to run the conversion faster and take up less memory.
+        buffer_mb: int (optional, defaults to 500MB)
+            Maximum amount of memory (in MB) to use per iteration of the internal DataChunkIterator.
+            Requires trace data in the RecordingExtractor to be a memmap object.
+        write_as: str (optional, defaults to 'raw')
+            Options: 'raw', 'lfp' or 'processed'
+        es_key: str (optional)
+            Key in metadata dictionary containing metadata info for the specific electrical series
+        icephys_experiment_type: str (optional)
+            Type of Icephys experiment. Allowed types are: 'voltage_clamp', 'current_clamp' and 'izero'.
+            If no value is passed, 'voltage_clamp' is used as default.
+        """
+        if (
+            HAVE_NDX_DANDI_ICEPHYS
+            and "ndx-dandi-icephys" in metadata
+            and "DandiIcephysMetadata" not in nwbfile.lab_meta_data
+        ):
+            nwbfile.add_lab_meta_data(DandiIcephysMetadata(**metadata["ndx-dandi-icephys"]))
+
+        for i, reader in enumerate(self.readers_list):
+            write_neo_to_nwb(
+                neo_reader=reader,
+                nwbfile=nwbfile,
+                metadata=metadata,
+                use_times=use_times,
+                write_as=write_as,
+                es_key=es_key,
+                save_path=save_path,
+                overwrite=overwrite if i == 0 else False,
+                icephys_experiment_type=metadata["Icephys"]["Sessions"][i]["icephys_experiment_type"],
+                stimulus_type=metadata["Icephys"]["Sessions"][i]["stimulus_type"],
```

Comment:
I notice that the `stimulus_type` and moreover the `Sessions` key are only filled in by the AbfInterface in this PR, with no mention or assertions that it is required by the metadata schema or as arguments of the `run_conversion` function itself.

We should generalize this in the base interface so it informs the user that this is somehow required, or takes on some specified type of default behavior if not.